### PR TITLE
Refactored graphical unit testing system, and implemented multi frames tests.

### PR DESCRIPTION
**Pull request by huguesdevimeux**
**NOTE : This PR aims to firstly refactor the testing system, I will modify the tests impacted by the new syntax (all the graphical tests) in a future PR.**

## Changelog / Overview

Refactored graphical unit testing system, and implemented multi frames tests.

## Explanation for Changes

I took inspiration of the testing system of matplotlib. 
The main thing is the introduction of the decorator `@frames_comparison`, that allow a better syntax for graphical unit test.
#### Before a graphical test looked like this : 
```py
class CreateTest(Scene):
    def construct(self):
        square = Square()
        self.play(Create(square))
# .. others tests .. 
@pytest.mark.parametrize("scene_to_test", get_scenes_to_test(__name__), indirect=False)
def test_scene(scene_to_test, tmpdir, show_diff):
    GraphicalUnitTester(scene_to_test[1], MODULE_NAME, tmpdir).test(show_diff=show_diff)
```
And, to set the control frames, some extra efforts were required, since you add to run this template with the current test 
```py
class YourClassTest(Scene):  # e.g. RoundedRectangleTest
    def construct(self):
        circle = Circle()
        self.play(Animation(circle))

set_test_scene(
    YourClassTest, "INSERT_MODULE_NAME"
)  # INSERT_MODULE_NAME can be e.g.  "geometry" or "movements"
```
It was not really a handful, and refrained a lot of people from writing tests. 
#### Now : 
The test looks like this : 
```py
@frames_comparison(test_name="CreateTest", module_name="creation", last_frame=False)
def test_create(scene):
    scene.play(Create(Square()))
    scene.wait()
``` 
To set the control frames, you just run 
```
pytest <your_test> --set_test
```

**One can test with multiple frames by setting `last_frame` argument to `False`. All frames produced by the test will be registered at a hard coded frame rate of 6.**
**This syntax also supports pytest marks and parametrization !** 

## Testing Status
I'm not sure how to test the setting of the control frames since this is pytest related, but if I find, I will do it. 

## Further comments : 
_Pfeww it was surprisingly hard_

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by hydrobeam** [comment]
What's the process for setting a test from a particular file?

I've tried  setting the control data for a dummy test from a file using of this:
```py
pytest mkt.py CreateTest --set_test
```
which produces an error:

```py
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --set_test
  inifile: C:\Users\laith\Documents\GitHub\hydro_manim\pyproject.toml
  rootdir: C:\Users\laith\Documents\GitHub\hydro_manim
```

**Comment by huguesdevimeux** [comment]
> What's the process for setting a test from a particular file?
> 
> I've tried setting the control data for a dummy test from a file at the root of the directory using of this:
> 
> ```python
> pytest mkt.py CreateTest --set_test
> ```
> 
> which produces an error:
> 
> ```python
> ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
> pytest: error: unrecognized arguments: --set_test
>   inifile: C:\Users\laith\Documents\GitHub\hydro_manim\pyproject.toml
>   rootdir: C:\Users\laith\Documents\GitHub\hydro_manim
> ```

Hmm looks like you're not in the right dir. i think you have to be in `manim/tests/` to be able to use `--set_test`

**Review by behackl — CHANGES_REQUESTED** [review]
I like this *a lot*, thank you very much for your efforts! I've left a bunch of comments and suggestions, pretty much all of them fairly minor.

And finally, there are some further comments I have in general:

- The private modules are fine, but I feel that `_FramesTester.py` is somewhat unpythonic -- what about renaming it to `_frames_tester.py`?
- Multi frame tests are pretty amazing! In order for us to use them somewhat liberally, I think that the render resolution should be adapted though. (I realize that making this possible requires a bit of restructuring, but I really do think that it is worth it: a multi frame test for a 2 seconds long animation produces a control data file  that is ~400kB large; with reduced resolution (say, 120x213) this could be brought down to ~30kB.
- It's possible that it is already the case before this PR, but on this branch pytest seems fairly sensitive to which directory it is called from -- I could only get it to work when calling it from the repository root, and got an error that the control data files were not found when calling it from somewhere else (e.g., from `tests/`).

All in all, great work! 👏

**Comment by huguesdevimeux** [comment]
> All in all, great work!
Thanks ! 
> The private modules are fine, but I feel that _FramesTester.py is somewhat unpythonic -- what about renaming it to _frames_tester.py? 

You're 100% right, my brain was in Java mode when I created the file and it didn't catch my attention. 

> Multi frame tests are pretty amazing! In order for us to use them somewhat liberally, I think that the render resolution should be adapted though. (I realize that making this possible requires a bit of restructuring, but I really do think that it is worth it: a multi frame test for a 2 seconds long animation produces a control data file that is ~400kB large; with reduced resolution (say, 120x213) this could be brought down to ~30kB.

Yes, a calculation must be made. I don't see any issue in reducing the resolution, and my initial idea was to encouraged people to use solo frame testing when possible (there are a lot of tests that only really test something in the last frame). 120x213 seems fine to me (the resolution can be very easily changed in the test config file)

> It's possible that it is already the case before this PR, but on this branch pytest seems fairly sensitive to which directory it is called from -- I could only get it to work when calling it from the repository root, and got an error that the control data files were not found when calling it from somewhere else (e.g., from tests/).

Arh, no, it's from this PR, I see the issue. I'm going to fix it ASAP. Thanks for the review !

File: tests/test_graphical_units/testing/_test_class_makers.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    base_scene: Type[Scene],
+    construct: Callable[[Scene], None],
+    test_renderer,
+) -> Type[Scene]:
+    class _TestedScene(base_scene):
+        def __init__(self, *args, **kwargs):
+            super().__init__(renderer=test_renderer, *args, **kwargs)
+
+        def construct(self):
+            construct(self)
```
[the remark this answers is not in the document]
Comment (huguesdevimeux) [reply tests/test_graphical_units/testing/_test_class_makers.py:20]: You're completly right, I will do that !

File: tests/test_graphical_units/testing/_test_class_makers.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+def _make_test_renderer_class(from_renderer):
+    # Just for heritance.
+    class _TestRenderer(from_renderer):
+        pass
+
+    return _TestRenderer
+
+
+class DummySceneFileWriter(SceneFileWriter):
```
[the remark this answers is not in the document]
Comment (huguesdevimeux) [reply tests/test_graphical_units/testing/_test_class_makers.py:37]: Hmmm if you're raising the fact that this class should not exist given that it is subclassed only once .. You're right. I has in mind another design when I wrote it. i will fix that.

File: tests/test_graphical_units/testing/frames_comparison.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    test_name : str
+        The name of the test.
+    module_name : str
+        The module which the test belongs to.
+    last_frame : bool, optional
+        whether the test should test the last frame, by default True.
+    renderer_class , optional
+        The base renderer to use (OpenGlRenderer/CairoRenderer), by default CairoRenderer
+    base_scene : [type], optional
+        The base class for the scene (ThreeDScene, etc.), by default Scene
```
[the remark this answers is not in the document]
Comment (huguesdevimeux) [reply tests/test_graphical_units/testing/frames_comparison.py:53]: I didn't know, thanks

File: tests/test_graphical_units/testing/_test_class_makers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (huguesdevimeux) [reply tests/test_graphical_units/testing/_test_class_makers.py:37]: On second thought, I like the way it is. It allows `_make_scene_file_writer_class` to be shorter. 

i'm open to debate though.

**Comment by huguesdevimeux** [comment]
@behackl I think I adressed all your changes !

File: tests/test_graphical_units/testing/frames_comparison.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+from ._frames_testers import _ControlDataWriter, _FramesTester
+from ._test_class_makers import (
+    DummySceneFileWriter,
+    _make_scene_file_writer_class,
+    _make_test_renderer_class,
+    _make_test_scene_class,
+)
+
+SCENE_PARAMETER_NAME = "scene"
+PATH_CONTROl_DATA = Path("tests", "control_data", "graphical_units_data").absolute()
```
Comment (behackl) [inline tests/test_graphical_units/testing/frames_comparison.py:25]: Ah, this absolute path here probably causes the path issue, right? Do you want to fix this still in the scope of this PR?

There also is a typo in the variable name. :-)

```suggestion
PATH_CONTROL_DATA = Path("tests", "control_data", "graphical_units_data").absolute()
```

File: tests/test_graphical_units/testing/frames_comparison.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            if last_frame:
+                frames_tester.check_frame(-1, scene_tested.renderer.get_frame())
+
+    return real_test
+
+
+def _get_control_data_path(
+    module_name: str, test_name: str, setting_test: bool
+) -> Path:
+    path = PATH_CONTROl_DATA / module_name
```
Comment (behackl) [inline tests/test_graphical_units/testing/frames_comparison.py:196]:
```suggestion
    path = PATH_CONTROL_DATA / module_name
```

**Review by behackl — APPROVED** [review]
One additional minor thing, but I'm happy otherwise.

I should have been more specific regarding multi-frame tests: what I meant was that I think it would be a good idea to have different resolutions for single-frame and multi-frame tests -- but this is certainly something that can be handled in a follow-up PR.

And same with the `DummySceneFileWriter` -- I was just asking about the name of the class, not really about the structure. I don't have a strong opinion about that though, keeping `DummySceneFileWriter` is fine for me.

**Comment by huguesdevimeux** [comment]
Arh, the path problem. I completely forgot it in like 5 minutes. I will fix it before merging the PR for sure !

**Comment by huguesdevimeux** [comment]
> I should have been more specific regarding multi-frame tests: what I meant was that I think it would be a good idea to have different resolutions for single-frame and multi-frame tests -- but this is certainly something that can be handled in a follow-up PR.

That's a neat idea, I'd prefer doing it **in this PR**. I will re-request a review when it's done :D

**Review by behackl — APPROVED** [review]

**Pull request merged**